File: authentication-service/src/main/java/com/michibaum/authentication_service/authentication/AuthenticationService.java
package com.michibaum.authentication_service.authentication;

import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.exceptions.JWTCreationException;
import com.michibaum.authentication_library.PublicKeyDto;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.security.KeyPair;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.util.HashMap;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class AuthenticationService {

    private final KeyPair keyPair;


    public PublicKeyDto getPublicKey() {
        return new PublicKeyDto(
                keyPair.getPublic().getAlgorithm(),
                keyPair.getPublic().getEncoded()
        );
    }

    public String generateJWS() {
        try {
            RSAPublicKey publicKey = (RSAPublicKey) keyPair.getPublic();
            RSAPrivateKey privateKey = (RSAPrivateKey) keyPair.getPrivate();
            Algorithm algorithm = Algorithm.RSA256(publicKey, privateKey);
            return JWT.create()
                    .withHeader(jwsHeaders())
                    .withIssuer("authentication-service")
                    .sign(algorithm);
        } catch (JWTCreationException exception){
            //Invalid Signing configuration / Couldn't convert Claims.
        }
        return null;
    }

    private Map<String, Object> jwsHeaders(){
        Map<String, Object> headers = new HashMap<String, Object>();
        headers.put("alg", keyPair.getPublic().getAlgorithm());
        return headers;
    }

}

File: authentication-service/src/main/kotlin/com/michibaum/authentication_service/authentication/AuthenticationService.kt
package com.michibaum.authentication_service.authentication

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import com.michibaum.authentication_library.PublicKeyDto
import org.springframework.stereotype.Service
import java.security.KeyPair
import java.security.interfaces.RSAPrivateKey
import java.security.interfaces.RSAPublicKey
import java.time.Instant

@Service
class AuthenticationService (
        private val keyPair: KeyPair,
        private val algorithm: Algorithm
) {

    val publicKey: PublicKeyDto
        get() = PublicKeyDto(
                keyPair.public.algorithm,
                keyPair.public.encoded
        )

    fun generateJWS(username: String): String? {
        return JWT.create()
                .withHeader(jwsHeaders())
                .withIssuer("authentication-service")
                .withSubject(username)
                .withExpiresAt(Instant.now().plusSeconds(60*60*8))
                .withIssuedAt(Instant.now())
                .sign(algorithm)
    }

    private fun jwsHeaders(): Map<String, String?> {
        return mapOf(Pair("alg", keyPair.public.algorithm))
    }
}